docs/enhancements/0001-worker-pods.md

```diff
@@ -0,0 +1,233 @@
+│    Pod     │  creates & owns  │ │ controller      ├───┼────────────────────┘
+│            │                  │ │                 │   │   watches &
+└──────┬─────┘                  │ └─────────────────┘   │   updates status
+       │                        │                       │
+       │                        │          KMM operator │
+       │                        └───────────────────────┘
+       │
+       │
+       │                        ┌──────────────────────┐
+       │                        │                      │
+       └────────────────────────► registry.example.org │
+                pulls from      │                      │
+                                └──────────────────────┘
+```
+
+#### The operator
+
+The operator runs two main controllers: the `Module` controller and the `NodeModuleState` controller.
+
+The `Module` controller watches both `Module` and `Node` resources to determine which kmods should be loaded on which
+`Node`.
+It creates one `NodeModuleState` resource per `Node` and `Module` that targets that node.  
+`NodeModuleState` resources represent the desired state of a `Module` on a node; they contain all the required data to
+pull the kmod image and load the module properly.  
+The `Module` controller watches updates to the `status` of `NodeModuleState` objects and aggregates that data into the
+`Module` `status`.
+
+The `NodeModuleState` controller watches `NodeModuleState` resources and determines when a worker Pod needs to be
+started on a node by looking at the difference between the desired state (spec) and the current one (status).
+When the Pod has completed, the controller reads its outcome and updates the `NodeModuleState` status accordingly.
+
+When updating the `NodeModuleState` status, the controller includes a timestamp corresponding to the last time a worker
+Pod ran on the node.
+This timestamp can be used to verify if a new worker Pod needs to run when the KMM operator restarts; if it is older
+than the last transition of the `Node`'s `Ready` condition, then the node may have rebooted and we potentially need to
+load the kernel modules.
+
+#### The worker Pods
+
+Worker Pods are created by the `NodeModuleState` controller.
+They run KMM code that reads the configuration file, pulls and extracts kmod OCI images and uses `modprobe` to load or
+unload kmods.  
+The configuration file is provided as an annotation to the Pod object and mounted in its filesystem via the [Downward
+API](https://kubernetes.io/docs/concepts/workloads/pods/downward-api/).
+
+##### Security
+
+Worker Pods run a privileged container but are granted zero privilege on the API Server.
+
+##### kmod images
+
+It is worth noting that worker Pods pull the kmod image from registries **into their filesystem**, not the node's.
+That filesystem disappears as soon as the Pod is terminated.  
+This means that node image caching mechanisms cannot be leveraged, and any Pod creation will entail data transfers from
```

Comment:
Can't we mount the node's cache to the worker pods?

Reply:
The nature and structure of the image cache varies on the container runtime used, and its storage driver.